google/cloud/documentai_toolbox/wrappers/document_revision.py

```diff
@@ -0,0 +1,502 @@
+            Required. The name of the output_bucket.
+
+        output_prefix (str):
+            Required. The prefix of the folder where files are excluding the bucket name.
+
+    Returns:
+        Tuple[str, list, list]:
+            document text, base document.proto, revision shards.
+
+    """
+    text = ""
+    base_doc = []
+    revisions_doc = []
+
+    storage_client = storage.Client()
+
+    blob_list = storage_client.list_blobs(output_bucket, prefix=output_prefix)
+    pb = documentai.Document.pb()
+    for blob in blob_list:
+        name = blob.name.split("/")[-1]
+        if blob.name.endswith(".dp.bp"):
+            blob_as_bytes = blob.download_as_bytes()
+            if re.search(r"^doc.dp.bp", name):
+                text = pb.FromString(blob_as_bytes).text
+            elif re.search(r"^pages_*.*", name):
+                base_doc.append(blob_as_bytes)
+            elif re.search(r"^rev_*.*", name):
+                revisions_doc.append(blob_as_bytes)
+
+    return text, base_doc, revisions_doc
+
+
+def _get_base_docproto(gcs_bucket_name, gcs_prefix) -> Tuple[list, list]:
+    r"""
+    Returns base document.proto and list of revision shards.
+
+    Args:
+        gcs_bucket_name (str):
+            Required. The gcs bucket.
+
+            Format: Given `gs://{bucket_name}/{optional_folder}/{operation_id}/` where `gcs_bucket_name={bucket_name}`.
+        gcs_prefix (str):
+            Required. The prefix to the location of the target folder.
+
+            Format: Given `gs://{bucket_name}/{optional_folder}/{target_folder}` where `gcs_prefix={optional_folder}/{target_folder}`.
+    Returns:
+        Tuple[list, list]:
+            base document.proto, revision shards.
+    """
+    base_shards = []
+    revision_shards = []
+
+    text, base_bytes, revision_bytes = _get_base_and_revision_bytes(
+        gcs_bucket_name, gcs_prefix
+    )
+    page_pb = documentai.Document.Page.pb()
+    pb = documentai.Document.pb()
+    for byte in base_bytes:
+        base_shards.append(
+            documentai.Document(pages=[page_pb.FromString(byte)], text=text)
+        )
+
+    for byte in revision_bytes:
+        revision_shards.append(pb.FromString(byte))
+
+    return base_shards, revision_shards
+
+
+def _get_revised_entites(
+    entities: List[documentai.Document.Entity],
+) -> Tuple[list, list]:
+    r"""Modifies the entities using the providence field in the entities.
+
+    Args:
+        entities (List[Entity]):
+            A list of documentai.Document.Entity.
+
+    Returns:
+        Tuple[list,list]:
+            modified entities, Entity history.
+    """
+    entities_array = []
+    history = []
+    for e in entities:
+        if (
+            e.provenance.type_ == _OP_TYPE.ADD
+            or e.provenance.type_ == _OP_TYPE.OPERATION_TYPE_UNSPECIFIED
+        ):
+            entities_array.append(Entity(documentai_entity=e))
+            history.append(
+                {
+                    "object": "Entity",
```

Comment:
If "object" is always "Entity" in these "history dictionaries", do we need "object"?

How general are these history dicts?  Do they deserve to be upgraded into a class or is it better to just keep them as such ad hoc Python dicts?  (for example, do we every need to pass histories into some functions as arguments?  should they carry methods and/or other attributes?)

Reply:
History can be used for more then just entities and history needs the `object` to work with CDW. Currently only entity revisions are supported but in the future other things can be supported. Also the structure of history is defined by the internal team so we can't really modify the structure. We can move the history into a class to make it easier to future proof this.